File: backend/src/main/java/tom/projects/monkeylog/mapper/WorkoutMapper.java
package tom.projects.monkeylog.mapper;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.ReportingPolicy;
import tom.projects.monkeylog.dto.workout.*;
import tom.projects.monkeylog.model.workout.ExerciseGroup;
import tom.projects.monkeylog.model.workout.ExerciseRow;
import tom.projects.monkeylog.model.workout.ExerciseRowField;
import tom.projects.monkeylog.model.workout.Workout;

import java.util.List;

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses={ExerciseMapper.class})
public interface WorkoutMapper {
    Workout workoutRequestToWorkout(WorkoutCreateRequest workoutRequest);

    WorkoutResponse workoutToWorkoutResponse(Workout workout);

    @Mapping(source=".", target="workout")
    WorkoutFullResponse workoutToFullWorkoutResponse(Workout workout);

    List<WorkoutFullResponse> workoutsToFullWorkoutResponses(List<Workout> workouts);

    ExerciseGroupResponse exerciseGroupToExerciseGroupResponse(ExerciseGroup exerciseGroup);

    ExerciseRowResponse exerciseRowToExerciseRowResponse(ExerciseRow exerciseRow);

    ExerciseRowFieldResponse exerciseRowFieldToExerciseRowFieldResponse(ExerciseRowField exerciseRowField);
}

File: backendkt/src/main/kotlin/tom/com/monkeylog/mapper/WorkoutMapper.kt
package tom.com.monkeylog.mapper

import org.mapstruct.Mapper
import org.mapstruct.Mapping
import org.mapstruct.ReportingPolicy
import tom.com.monkeylog.dto.workout.*
import tom.com.monkeylog.model.workout.ExerciseGroup
import tom.com.monkeylog.model.workout.ExerciseRow
import tom.com.monkeylog.model.workout.ExerciseRowField
import tom.com.monkeylog.model.workout.Workout

@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE, uses = [ExerciseMapper::class])
interface WorkoutMapper {
    fun workoutRequestToWorkout(workoutRequest: WorkoutCreateRequest): Workout
    fun workoutToWorkoutResponse(workout: Workout): WorkoutResponse

    @Mapping(source = ".", target = "workout")
    fun workoutToFullWorkoutResponse(workout: Workout): WorkoutFullResponse
    fun workoutsToFullWorkoutResponses(workouts: List<Workout>): List<WorkoutFullResponse>
    fun exerciseGroupToExerciseGroupResponse(exerciseGroup: ExerciseGroup): ExerciseGroupResponse
    fun exerciseRowToExerciseRowResponse(exerciseRow: ExerciseRow): ExerciseRowResponse
    fun exerciseRowFieldToExerciseRowFieldResponse(exerciseRowField: ExerciseRowField): ExerciseRowFieldResponse
}
